### PR TITLE
Update - Improve repo JavaScript tooling

### PR DESCRIPTION
**Pull request by rosborne132**
## Summary of changes

Describe the reason for the change.
- Add tooling for auto formatting of any new JavaScript code written into the project. We currently have linting in the project but it doesn't format the code. Added Preitter, lint-staged and husky to lint and format staged code for commits.
- Added a pull request template.

## Type of change

Please mark the relevant options with an 'x'.

~~- [ ] Bug fix (non-breaking change which fixes an issue)~~
~~- [ ] New feature (non-breaking change which adds functionality)~~
~~- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)~~
~~- [ ] Backfills tests~~
~~- [ ] This change requires a documentation update~~
- [x] This change provides tooling improvements

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "tabWidth": 4,
```
Comment (rosborne132) [inline .prettierrc:2]: These rules we based of our eslint rules.

File: package.json [hunk 2]
```diff
@@ -7,15 +7,25 @@
...
     "eslint": "^8.34.0",
+    "husky": "^8.0.3",
     "jest": "^29.3.1",
+    "lint-staged": "^13.2.1",
+    "prettier": "^2.8.7",
     "tsembind": "github:ted537/tsembind#988aea8",
     "typescript": "^5.0.3"
+  },
+  "lint-staged": {
+    "tests/**/*.js": [
```
Comment (rosborne132) [inline package.json:26]: I've limited this to the tests folder for now.

**Comment by JeanChristopheMorinPerso** [comment]
Hey @rosborne132, thanks for helping me out!

I was initially (and still am) happy with ESLint "basic" formatting (`eslint --fix` in this case), but if you think there is an advantage of using prettier on top of it, then why not. Should we remove this block: https://github.com/JeanChristopheMorinPerso/OpenTimelineIO-JS-Bindings/blob/main/.eslintrc.json#L19?

As for pre-commits, my personal opinion is that they are not super user friendly IMO. And they tend to make the commit experience pretty bad (it's usually slow and quite frustrating to not be able to commit temporary commits).

I prefer to catch linting and formatting issues in CI. In fact, ESLint already runs in CI, see https://github.com/JeanChristopheMorinPerso/OpenTimelineIO-JS-Bindings/blob/main/.github/workflows/lint.yaml.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+## Type of change
+
+Please mark the relevant options with an 'x'.
+
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
+- [ ] Backfills tests
+- [ ] This change requires a documentation update
+- [ ] This change provides tooling improvements
```
Comment (JeanChristopheMorinPerso) [inline .github/PULL_REQUEST_TEMPLATE.md]: How about:
```suggestion
## Summary of changes

Fixes #123.

Describe the reason for the change.

Add a list of changes, and note any that might need special attention during review. Help us understand what this pull request is changing and why it's needed.
```
?

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]
Hey @rosborne132, thanks for helping me out!

I was initially (and still am) happy with ESLint "basic" formatting (eslint --fix in this case), but if you think there is an advantage of using prettier on top of it, then why not. Should we remove this block: https://github.com/JeanChristopheMorinPerso/OpenTimelineIO-JS-Bindings/blob/main/.eslintrc.json#L19?

As for pre-commits, my personal opinion is that they are not super user friendly IMO. And they tend to make the commit experience pretty bad (it's usually slow and quite frustrating to not be able to commit temporary commits).

I prefer to catch linting and formatting issues in CI. In fact, ESLint already runs in CI, see https://github.com/JeanChristopheMorinPerso/OpenTimelineIO-JS-Bindings/blob/main/.github/workflows/lint.yaml.

**Comment by rosborne132** [comment]
@JeanChristopheMorinPerso Thanks for the feedback. When it comes to formatting code it is better practice to have the code formatted on the local machine. It would be much faster to format code through a pre-commit compared to waiting for the results in CI. We can continue to use CI to lint continue based on the recommend rules provided by eslint but I suggest to leave the formatting of the code to the local machine with the prettier rules outlined in the PR.

File: .eslintrc.json [hunk 4]
```diff
@@ -16,18 +16,5 @@
         "emsdk/**/*",
         "repros/**/*"
     ],
-    "rules": {
```
Comment (rosborne132) [inline .eslintrc.json:19]: Removing rules sense they are style related. Style related issue will be addressed through pritter.

**Comment by JeanChristopheMorinPerso** [comment]
I agree with the fact that code formatting should run locally. And that's what I do. But I do that through my IDE. If we are to add a more magical code formatting solution, I would prefer something that doesn't require local pre-commits. I really think creating a commit should be almost instantaneous (which means no pre-commits that run code formatters).

Also, note that this repo will eventually be moved to the OTIO org, where no repos use pre-commits.

You can also setup a local pre-commit in your local fork if you want automatic formatting when you commit. This would be perfectly valid and wouldn't force anyone to also use them. (Anything in the `.git/hooks` folder is not tracked by git.

**Comment by rosborne132** [comment]
Closing PR and carrying over some changes into this [PR](https://github.com/JeanChristopheMorinPerso/OpenTimelineIO-JS-Bindings/pull/9)

**Pull request closed without merging**